eth/protocols/snap/sync.go

```diff
@@ -2201,11 +2201,16 @@ func (s *Syncer) processStorageResponse(res *storageResponse) {
 			// If the chunk's root is an overflown but full delivery,
 			// clear the heal request.
 			accountHash := res.accounts[len(res.accounts)-1]
-			if root == res.subTask.root && rawdb.HasStorageTrieNode(s.db, accountHash, nil, root) {
-				for i, account := range res.mainTask.res.hashes {
-					if account == accountHash {
-						res.mainTask.needHeal[i] = false
-						skipStorageHealingGauge.Inc(1)
+			if root == res.subTask.root {
+				// Ensure the root node with particular hash is present in disk
+				// before clearing.
+				blob := rawdb.ReadStorageTrieNode(s.db, accountHash, nil)
+				if len(blob) != 0 && crypto.Keccak256Hash(blob) == root {
```

Comment:
Why did you change from `HasStorageTrieNode` into `ReadStorageTrieNode`? Seems like you did it to avoid doing a keccak on `len(blob)==0`?  perhaps then that should be internalized into the rawdb methods?

Reply:
Originally the `HasStorageTrieNode` will compare if the found node is hash-matched with the given hash. In this pull request, the hash comparison in `HasStorageTrieNode` is dropped.

We use `ReadStorageTrieNode` here to manually compare the hash. The new code should be logic equivalent with the old one.